src/empirical.jl

```diff
@@ -1,68 +1,142 @@
+    return res
 end
 
 """
-    ecdf(X; weights::AbstractWeights)
+    ecdf(X::RealVector; weights::AbstractWeights=nothing, interpolate::Bool=false) -> ECDF
+
+Construct an *empirical cumulative distribution function* (ECDF) based on a vector of samples (`X`).
+Returns a callable [`ECDF`](@ref) object.
 
-Return an empirical cumulative distribution function (ECDF) based on a vector of samples
-given in `X`. Optionally providing `weights` returns a weighted ECDF.
+# Example
 
-Note: this function that returns a callable composite type, which can then be applied to
-evaluate CDF values on other samples.
+```jldoctest
+julia> using StatsBase
 
-`extrema`, `minimum`, and `maximum` are supported to for obtaining the range over which
-function is inside the interval ``(0,1)``; the function is defined for the whole real line.
+julia> X = 1:15
+1:15
+
+julia> xcdf = ecdf(X)
+ECDF{Int64,Float64,false}(15 values)
+
+julia> xcdf(9)
+0.6
+```
+
+## Keyword arguments
+* `weights`: optional weights for each `X` sample
+* `interpolate`: when enabled, `ECDF(y)` will linearly interpolate the cumulative density between
+  the neighboring `X` samples (``x_i \\leq y < x_{i+1}``),
+  otherwise (by default) `ECDF(x[i])` is returned
+
+## Notes
+The returned function is defined for the whole real line.
+Use [`extrema`](@ref), [`minimum`](@ref), and [`maximum`](@ref) to obtain the range
+over which the ECDF is inside the interval ``(0, 1)``.
 """
-function ecdf(X::RealVector; weights::AbstractVector{<:Real}=Weights(Float64[]))
+function ecdf(X::RealVector;
+              weights::Union{Nothing, RealVector}=nothing,
+              interpolate::Bool=false)
     any(isnan, X) && throw(ArgumentError("ecdf can not include NaN values"))
-    isempty(weights) || length(X) == length(weights) || throw(ArgumentError("data and weight vectors must be the same size," *
-        "got $(length(X)) and $(length(weights))"))
+    evenweights = isnothing(weights) || isempty(weights)
+    evenweights || (length(X) == length(weights)) ||
+        throw(ArgumentError("data and weight vectors must be the same size, " *
+                            "got $(length(X)) and $(length(weights))"))
+    T = eltype(X)
+    W0 = evenweights ? Int : eltype(weights)
+    W = isnothing(weights) ? Float64 : eltype(one(W0)/sum(weights))
+
+    wsum = evenweights ? length(X) : sum(weights)
     ord = sortperm(X)
-    ECDF(X[ord], isempty(weights) ? weights : Weights(weights[ord]))
+
+    sorted_vals = sizehint!(Vector{Tuple{T, W, W, W}}(), length(X))
+    isempty(X) && return ECDF{T, W, interpolate}(sorted_vals)
+
+    valprev = val = X[first(ord)]
+    wsumprev = zero(W0)
+    valw = zero(W0)
+
+    push_valprev!() = push!(sorted_vals, (valprev, min(wsumprev/wsum, one(W)),
+                                          inv(val - valprev), valw/wsum))
+
+    @inbounds for i in ord
+        valnew = X[i]
+        if (val != valnew) || (i == last(ord))
+            (wsumprev > 0) && push_valprev!()
```

Comment:
Doesn't this anonymous function trigger boxing of these variables, making the loop quite slow?